### ISSUE TITLE
Intermittent failure of twisted.test.test_twistd.AppProfilingTestCase.test_profileSaveStats on Windows

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3222]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3222|
|Type|defect|
|Created|2008-05-04 22:44:57Z|
|Branch|https://github.com/twisted/twisted/tree/binary-profile-files-3222|

Seen on BuildBot:

```
[FAIL]: twisted.test.test_twistd.AppProfilingTestCase.test_profileSaveStats

Traceback (most recent call last):
  File "c:\twistedbuildbot24\WXP32-full2.5-scmikes-select\Twisted\twisted\test\test_twistd.py", line 453, in test_profileSaveStats
    self.assertIn("DummyReactor.run", data)
twisted.trial.unittest.FailTest: 'DummyReactor.run' not in '{(\x03\x00\x00\x00t\x00\x00\x00\x00i\x00\x00\x00\x00s\n\x00\x00\x00setprofile(\x05\x00\x00\x00i\x01\x00\x00\x00i\x01\x00\x00\x00g\x00\x80\x81j0'
```

<details><summary>Searchable metadata</summary>

```
trac-id__3222 3222
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_binary_profile_files_3222 branches/binary-profile-files-3222
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1209941097000000 1209941097000000
changetime__1221245511000000 1221245511000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|z3p set status to `reopened`|
|-|-|

(In [24480]) Revert r24477

Reopens [#3222](https://github.com/twisted/twisted/issues/3222)

There was a Python 2.3 incompatibility with the previous patch.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|z3p commented|
|-|-|

Wrong bug number in the SVN checkin.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [24800]) Branching to 'binary-profile-files-3222'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

Some background: the test would fail because the test itself loaded back the stats dump using a file open in text mode; every once in a while, the stats data would happen to marshal with a !^Z, which means EOF for text files on Windows, so most of the data wouldn't be read back in by the test.

So, hmm.  I suppose this is fixed.  I changed the tests to make more realistic (I hope) assertions about the output of the profile tests.  The tests actually load the stats data and then re-dump it as text and make assertions about the text.  If the tests pass now, it should mean that the stats dump is meaningful and correct.

The tests pass quite a lot now on Windows (2000+ iterations of success), whereas before about 200 iterations was enough to trigger the failure.  I didn't change any implementation code, though - the stats are still dumped to files opened in text mode.  I really feel like that should be wrong, but the tests don't fail...

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

Fine, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [24809]) Merge binary-profile-files-3222

Author: exarkun
Reviewer: therve
Fixes: [#3222](https://github.com/twisted/twisted/issues/3222)

Change the Application profiler tests so that they make
assertions by loading the dumped stats with a suitable
API (`pstats.Stats` or `hotshot.stats.load`) and then
reformatting the data as text, rather than making
assertions about the opaque binary data directly.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
